In [1]:
import pandas as pd
import numpy as np

In [2]:
df = pd.read_csv('customer_supermarket.csv', sep='\t', index_col=0)

In [3]:
df.head(10)

BasketID      BasketDate  Sale  CustomerID CustomerCountry  ProdID  \
0   536365  01/12/10 08:26  2,55     17850.0  United Kingdom  85123A   
1   536365  01/12/10 08:26  3,39     17850.0  United Kingdom   71053   
2   536365  01/12/10 08:26  2,75     17850.0  United Kingdom  84406B   
3   536365  01/12/10 08:26  3,39     17850.0  United Kingdom  84029G   
4   536365  01/12/10 08:26  3,39     17850.0  United Kingdom  84029E   
5   536365  01/12/10 08:26  7,65     17850.0  United Kingdom   22752   
6   536365  01/12/10 08:26  4,25     17850.0  United Kingdom   21730   
7   536366  01/12/10 08:28  1,85     17850.0  United Kingdom   22633   
8   536366  01/12/10 08:28  1,85     17850.0  United Kingdom   22632   
9   536368  01/12/10 08:34  4,25     13047.0  United Kingdom   22960   

                             ProdDescr  Qta  
0   WHITE HANGING HEART T-LIGHT HOLDER    6  
1                  WHITE METAL LANTERN    6  
2       CREAM CUPID HEARTS COAT HANGER    8  
3  KNITTED UNION FLAG HOT WATER BOTTLE    6  
4       RED WOOLLY HOTTIE WHITE HEART.    6  
5         SET 7 BABUSHKA NESTING BOXES    2  
6    GLASS STAR FROSTED T-LIGHT HOLDER    6  
7               HAND WARMER UNION JACK    6  
8            HAND WARMER RED POLKA DOT    6  
9             JAM MAKING SET WITH JARS    6

In [4]:
df.tail(10)

BasketID      BasketDate  Sale  CustomerID CustomerCountry ProdID  \
541900   581587  09/12/11 12:50  3,75     12680.0          France  22730   
541901   581587  09/12/11 12:50  1,95     12680.0          France  22367   
541902   581587  09/12/11 12:50  1,95     12680.0          France  22629   
541903   581587  09/12/11 12:50  4,15     12680.0          France  23256   
541904   581587  09/12/11 12:50  0,85     12680.0          France  22613   
541905   581587  09/12/11 12:50   2,1     12680.0          France  22899   
541906   581587  09/12/11 12:50  4,15     12680.0          France  23254   
541907   581587  09/12/11 12:50  4,15     12680.0          France  23255   
541908   581587  09/12/11 12:50  4,95     12680.0          France  22138   
541909   581587  09/12/11 12:50    18     12680.0          France   POST   

                              ProdDescr  Qta  
541900       ALARM CLOCK BAKELIKE IVORY    4  
541901  CHILDRENS APRON SPACEBOY DESIGN    8  
541902              SPACEBOY LUNCH BOX    12  
541903      CHILDRENS CUTLERY SPACEBOY     4  
541904      PACK OF 20 SPACEBOY NAPKINS   12  
541905     CHILDREN'S APRON DOLLY GIRL     6  
541906    CHILDRENS CUTLERY DOLLY GIRL     4  
541907  CHILDRENS CUTLERY CIRCUS PARADE    4  
541908    BAKING SET 9 PIECE RETROSPOT     3  
541909                          POSTAGE    1

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 471910 entries, 0 to 541909
Data columns (total 8 columns):
 #   Column           Non-Null Count   Dtype  
---  ------           --------------   -----  
 0   BasketID         471910 non-null  object 
 1   BasketDate       471910 non-null  object 
 2   Sale             471910 non-null  object 
 3   CustomerID       406830 non-null  float64
 4   CustomerCountry  471910 non-null  object 
 5   ProdID           471910 non-null  object 
 6   ProdDescr        471157 non-null  object 
 7   Qta              471910 non-null  int64  
dtypes: float64(1), int64(1), object(6)
memory usage: 32.4+ MB


Something wrong with the indeces. Indeces from 0 to 541909 but only 471910 entries.

In [6]:
df.shape

(471910, 8)

Read again the dataset with implicit indexing and delete the first column (which conteins the wrong indeces)

In [7]:
df = pd.read_csv('customer_supermarket.csv', sep='\t')
df.head()

Unnamed: 0 BasketID      BasketDate  Sale  CustomerID CustomerCountry  \
0           0   536365  01/12/10 08:26  2,55     17850.0  United Kingdom   
1           1   536365  01/12/10 08:26  3,39     17850.0  United Kingdom   
2           2   536365  01/12/10 08:26  2,75     17850.0  United Kingdom   
3           3   536365  01/12/10 08:26  3,39     17850.0  United Kingdom   
4           4   536365  01/12/10 08:26  3,39     17850.0  United Kingdom   

   ProdID                            ProdDescr  Qta  
0  85123A   WHITE HANGING HEART T-LIGHT HOLDER    6  
1   71053                  WHITE METAL LANTERN    6  
2  84406B       CREAM CUPID HEARTS COAT HANGER    8  
3  84029G  KNITTED UNION FLAG HOT WATER BOTTLE    6  
4  84029E       RED WOOLLY HOTTIE WHITE HEART.    6

In [8]:
df.drop(columns=['Unnamed: 0'], inplace=True)
df.head()

BasketID      BasketDate  Sale  CustomerID CustomerCountry  ProdID  \
0   536365  01/12/10 08:26  2,55     17850.0  United Kingdom  85123A   
1   536365  01/12/10 08:26  3,39     17850.0  United Kingdom   71053   
2   536365  01/12/10 08:26  2,75     17850.0  United Kingdom  84406B   
3   536365  01/12/10 08:26  3,39     17850.0  United Kingdom  84029G   
4   536365  01/12/10 08:26  3,39     17850.0  United Kingdom  84029E   

                             ProdDescr  Qta  
0   WHITE HANGING HEART T-LIGHT HOLDER    6  
1                  WHITE METAL LANTERN    6  
2       CREAM CUPID HEARTS COAT HANGER    8  
3  KNITTED UNION FLAG HOT WATER BOTTLE    6  
4       RED WOOLLY HOTTIE WHITE HEART.    6

In [9]:
df.tail()

BasketID      BasketDate  Sale  CustomerID CustomerCountry ProdID  \
471905   581587  09/12/11 12:50   2,1     12680.0          France  22899   
471906   581587  09/12/11 12:50  4,15     12680.0          France  23254   
471907   581587  09/12/11 12:50  4,15     12680.0          France  23255   
471908   581587  09/12/11 12:50  4,95     12680.0          France  22138   
471909   581587  09/12/11 12:50    18     12680.0          France   POST   

                              ProdDescr  Qta  
471905     CHILDREN'S APRON DOLLY GIRL     6  
471906    CHILDRENS CUTLERY DOLLY GIRL     4  
471907  CHILDRENS CUTLERY CIRCUS PARADE    4  
471908    BAKING SET 9 PIECE RETROSPOT     3  
471909                          POSTAGE    1

In [10]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 471910 entries, 0 to 471909
Data columns (total 8 columns):
 #   Column           Non-Null Count   Dtype  
---  ------           --------------   -----  
 0   BasketID         471910 non-null  object 
 1   BasketDate       471910 non-null  object 
 2   Sale             471910 non-null  object 
 3   CustomerID       406830 non-null  float64
 4   CustomerCountry  471910 non-null  object 
 5   ProdID           471910 non-null  object 
 6   ProdDescr        471157 non-null  object 
 7   Qta              471910 non-null  int64  
dtypes: float64(1), int64(1), object(6)
memory usage: 28.8+ MB


Now entries are indexed well. 
Why is Sale of type object? Why is CustomerID of type float?

In [11]:
df_nullCustID = df[df.CustomerID.isnull()]
df_nullCustID

BasketID      BasketDate     Sale  CustomerID CustomerCountry  ProdID  \
1442     536544  01/12/10 14:32     2,51         NaN  United Kingdom   21774   
1443     536544  01/12/10 14:32     1,66         NaN  United Kingdom   21787   
1444     536544  01/12/10 14:32     1,66         NaN  United Kingdom   21790   
1445     536544  01/12/10 14:32     2,51         NaN  United Kingdom   21791   
1446     536544  01/12/10 14:32     0,43         NaN  United Kingdom   21802   
...         ...             ...      ...         ...             ...     ...   
471536   581498  09/12/11 10:26     4,13         NaN  United Kingdom   85038   
471537   581498  09/12/11 10:26     3,29         NaN  United Kingdom  85049a   
471538   581498  09/12/11 10:26     4,96         NaN  United Kingdom   85150   
471539   581498  09/12/11 10:26    10,79         NaN  United Kingdom   85174   
471540   581498  09/12/11 10:26  1714,17         NaN  United Kingdom     DOT   

                                 ProdDescr  Qta  
1442       DECORATIVE CATS BATHROOM BOTTLE    2  
1443                 RAIN PONCHO RETROSPOT    2  
1444                    VINTAGE SNAP CARDS    9  
1445    VINTAGE HEADS AND TAILS CARD GAME     2  
1446       CHRISTMAS TREE HEART DECORATION    9  
...                                    ...  ...  
471536     6 CHOCOLATE LOVE HEART T-LIGHTS    1  
471537       TRADITIONAL CHRISTMAS RIBBONS    5  
471538       LADIES & GENTLEMEN METAL SIGN    1  
471539                   S/4 CACTI CANDLES    1  
471540                      DOTCOM POSTAGE    1  

[65080 rows x 8 columns]

In [12]:
df_nullCustID['CustomerCountry'].value_counts()

United Kingdom    64383
EIRE                339
Hong Kong           138
Unspecified          96
Switzerland          55
France               33
Israel               19
Portugal             15
Bahrain               2
Name: CustomerCountry, dtype: int64

In [13]:
df_nullDesc = df[df.ProdDescr.isnull()]
df_nullDesc[df_nullDesc.CustomerID.isnull()]

BasketID      BasketDate Sale  CustomerID CustomerCountry  ProdID  \
1729     536546  01/12/10 14:33    0         NaN  United Kingdom   22145   
1765     536550  01/12/10 14:34    0         NaN  United Kingdom   85044   
1766     536553  01/12/10 14:35    0         NaN  United Kingdom   37461   
3738     536764  02/12/10 14:42    0         NaN  United Kingdom  84952C   
3739     536765  02/12/10 14:43    0         NaN  United Kingdom  84952C   
...         ...             ...  ...         ...             ...     ...   
455480   580381  02/12/11 17:58    0         NaN  United Kingdom   21758   
467099   581199  07/12/11 18:26    0         NaN  United Kingdom   84581   
467102   581203  07/12/11 18:31    0         NaN  United Kingdom   23406   
467104   581209  07/12/11 18:35    0         NaN  United Kingdom   21620   
467988   581234  08/12/11 10:33    0         NaN  United Kingdom   72817   

       ProdDescr  Qta  
1729         NaN    1  
1765         NaN    1  
1766         NaN    3  
3738         NaN  -38  
3739         NaN   19  
...          ...  ...  
455480       NaN   -9  
467099       NaN   -2  
467102       NaN   15  
467104       NaN    6  
467988       NaN   27  

[753 rows x 8 columns]

All rows that have null prod. description have null customerID too.
There are also negative quantities.

In [14]:
df_nullDesc['Sale'].value_counts()

0    753
Name: Sale, dtype: int64

All rows that have null prod. description have value 0 in Sale. 

In [15]:
df_nullDesc['CustomerCountry'].value_counts()

United Kingdom    753
Name: CustomerCountry, dtype: int64

All rows that have null prod. description have value United Kingdom in CustomerCountry.


In [16]:
df['BasketID'].value_counts()

576339    542
573585    535
579196    533
580727    529
578270    442
         ... 
571297      1
546151      1
555489      1
566325      1
537461      1
Name: BasketID, Length: 24627, dtype: int64

In [17]:
df['BasketDate'].value_counts()

14/11/11 15:27    543
31/10/11 14:41    535
28/11/11 15:54    534
05/12/11 17:17    530
23/11/11 13:39    444
                 ... 
16/06/11 13:29      1
03/06/11 10:36      1
28/03/11 15:16      1
16/05/11 14:17      1
19/09/11 16:37      1
Name: BasketDate, Length: 22428, dtype: int64

In [27]:
sale = df['Sale'].value_counts()
sale

1,25      48519
1,65      37834
2,95      27492
0,85      27407
0,42      23230
          ...  
218,08        1
282,88        1
275,6         1
1,32          1
593,15        1
Name: Sale, Length: 1146, dtype: int64

In [33]:
sale['0']

1279

'Sale' attribute has value 0 in 1279 rows.

In [19]:
df['CustomerID'].value_counts()

17841.0    7983
14911.0    5903
14096.0    5128
12748.0    4642
14606.0    2782
           ... 
12943.0       1
13135.0       1
18068.0       1
17347.0       1
16093.0       1
Name: CustomerID, Length: 4372, dtype: int64

In [20]:
df['CustomerCountry'].value_counts()

United Kingdom          426261
Germany                   9495
France                    8525
EIRE                      7824
Spain                     2533
Netherlands               2371
Belgium                   2069
Switzerland               1932
Portugal                  1495
Australia                 1259
Norway                    1086
Italy                      803
Channel Islands            758
Finland                    695
Cyprus                     622
Sweden                     462
Austria                    401
Denmark                    389
Japan                      358
Poland                     341
Unspecified                340
USA                        291
Israel                     269
Singapore                  229
Iceland                    182
Canada                     151
Greece                     146
Hong Kong                  138
Malta                      127
United Arab Emirates        68
European Community          61
RSA                         58
Lebanon 

There is 'Unspecified' as country name.

In [21]:
df['ProdID'].value_counts()

85123A    2191
22423     2042
85099B    1917
47566     1569
20725     1492
          ... 
90037D       1
90074        1
85035b       1
72789        1
62095B       1
Name: ProdID, Length: 3953, dtype: int64

In [22]:
df['ProdDescr'].value_counts()

WHITE HANGING HEART T-LIGHT HOLDER    2212
REGENCY CAKESTAND 3 TIER              2042
JUMBO BAG RED RETROSPOT               1917
PARTY BUNTING                         1569
LUNCH BAG RED RETROSPOT               1491
                                      ... 
WRAP BLUE RUSSIAN FOLKART                1
Sale error                               1
GARLAND, MAGIC GARDEN 1.8M               1
MUMMY MOUSE RED GINGHAM RIBBON           1
M/COLOUR POM-POM CURTAIN                 1
Name: ProdDescr, Length: 4097, dtype: int64

In [34]:
qta = df['Qta'].value_counts()
qta

 1        109408
 2         69537
 12        60508
 6         39152
 4         35165
           ...  
 80995         1
 374           1
-1100          1
 3906          1
 2700          1
Name: Qta, Length: 597, dtype: int64

There are negative quantities.

In [35]:
df.describe()

CustomerID            Qta
count  406830.000000  471910.000000
mean    15287.684160      10.716533
std      1713.603074     231.355136
min     12346.000000  -80995.000000
25%     13953.000000       1.000000
50%     15152.000000       4.000000
75%     16791.000000      12.000000
max     18287.000000   80995.000000

'Sale' is missing from the description because it is an object...why? 

As previously noted, there are negative quantities.

Negative quantities could be replaced with their absolute value.